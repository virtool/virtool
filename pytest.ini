[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
