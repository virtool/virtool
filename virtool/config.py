import asyncio
import json
import logging
import os
import sys

import click
import psutil
import uvloop

import virtool.app
import virtool.db.mongo
import virtool.db.utils
import virtool.jobs.classes
import virtool.jobs.job
import virtool.jobs.run
import virtool.jobs.runner
import virtool.logs
import virtool.redis
import virtool.utils

logger = logging.getLogger(__name__)

CONFIG_PATH = os.path.join(sys.path[0], "config.json")

JOB_LIMIT_KEYS = (
    "lg_proc",
    "lg_mem",
    "sm_proc",
    "sm_mem"
)


def create_default_map():
    try:
        with open("./config.json", "r") as f:
            return json.load(f)
    except FileNotFoundError:
        return None


def validate_limits(config):
    cpu_count = psutil.cpu_count()
    mem_total = psutil.virtual_memory().total

    proc = int(config["proc"])
    mem = int(config["mem"])

    fatal = False

    if proc > cpu_count:
        fatal = True
        logger.fatal(f"Configured proc limit ({proc}) exceeds host CPU count ({cpu_count})")

    in_bytes = mem * 1024 * 1024 * 1024

    if in_bytes > mem_total:
        fatal = True
        logger.fatal(f"Configured mem limit ({in_bytes}) exceeds host memory ({mem_total})")

    for job_limit_key in JOB_LIMIT_KEYS:
        resource_key = job_limit_key.split("_")[1]

        job_limit = int(config[job_limit_key])
        host_limit = int(config[resource_key])

        if job_limit > host_limit:
            fatal = True
            logger.fatal(
                f"Configured {job_limit_key} ({job_limit}) exceeds instance {resource_key} limit ({host_limit})"
            )

    if fatal:
        sys.exit(1)

    return cpu_count, mem_total


def entry():
    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
    default_map = create_default_map()
    cli(auto_envvar_prefix="VT", default_map=default_map)


@click.group()
@click.option(
    "--data-path",
    default="data",
    help="The path to the application data directory",
    type=click.Path()
)
@click.option(
    "--db-connection-string",
    default="mongodb://localhost:27017",
    help="The MongoDB connection string",
    type=str
)
@click.option(
    "--db-name",
    default="virtool",
    help="The MongoDB database name",
    type=str
)
@click.option(
    "--dev",
    help="Run in development mode",
    is_flag=True
)
@click.option(
    "--force-version",
    help="Make the instance think it is a different version",
    type=str
)
@click.option(
    "--no-sentry",
    help="Disable Sentry error reporting",
    is_flag=True
)
@click.option(
    "--postgres-connection-string",
    help="The PostgreSQL connection string (must begin with 'postgresql+asyncpg://')",
    type=str,
    default="postgresql+asyncpg://virtool:virtool@localhost/virtool"
)
@click.option(
    "--proxy",
    help="The address for an internet proxy to connect through",
    type=str
)
@click.option(
    "--redis-connection-string",
    help="The Redis connection string",
    type=str,
    required=True,
    default="redis://localhost:6379"
)
@click.option(
    "--verbose",
    help="Log debug messages",
    is_flag=True
)
@click.pass_context
def cli(ctx, data_path, db_connection_string, db_name, dev, force_version, no_sentry, proxy, postgres_connection_string,
        redis_connection_string, verbose):
    ctx.ensure_object(dict)
    ctx.obj.update({
        "data_path": data_path,
        "db_connection_string": db_connection_string,
        "db_name": db_name,
        "dev": dev,
        "force_version": force_version,
        "no_sentry": no_sentry,
        "proxy": proxy,
        "postgres_connection_string": postgres_connection_string,
        "redis_connection_string": redis_connection_string,
        "verbose": verbose
    })


@cli.command("server", help="Start a Virtool API and websocket server")
@click.option(
    "--host",
    default="localhost",
    help="The host to listen on",
    type=str
)
@click.option(
    "--port",
    default=9950,
    help="The port to listen on",
    type=int
)
@click.option(
    "--no-check-db",
    help="Start without checking and repairing database",
    is_flag=True
)
@click.option(
    "--no-check-files",
    help="Start without ensuring data directory is valid",
    is_flag=True
)
@click.option(
    "--no-client",
    help="Start without serving client files",
    is_flag=True
)
@click.option(
    "--no-fetching",
    help="Start with automatic fetching disabled",
    is_flag=True
)
@click.pass_context
def start_server(ctx, host, port, no_check_db, no_check_files, no_client, no_fetching):
    virtool.logs.configure_server(ctx.obj["dev"], ctx.obj["verbose"])

    config = {
        **ctx.obj,
        "host": host,
        "port": port,
        "no_check_db": no_check_db,
        "no_check_files": no_check_files,
        "no_client": no_client,
        "no_fetching": no_fetching
    }

    logger.info("Starting in server mode")
    asyncio.get_event_loop().run_until_complete(virtool.app.run_app(config))


@cli.command("runner", help="Start a runner that runs one job at a time")
@click.option(
    "--job-list", "-l",
    default=["jobs_lg", "jobs_sm"],
    help="A Redis list key to pull job IDs from",
    multiple=True,
    type=str
)
@click.option(
    "--mem",
    default=1,
    help="The maximum memory (GB) the instance may use",
    type=int
)
@click.option(
    "--proc",
    default=1,
    help="The maximum number of processes the instance can use",
    type=int
)
@click.option(
    "--temp-path",
    type=click.Path(),
    help="The path to local directory for temporary files"
)
@click.pass_context
def start_runner(ctx, job_list, mem, proc, temp_path):
    virtool.logs.configure_runner(ctx.obj["dev"], ctx.obj["verbose"])

    config = {
        **ctx.obj,
        "job_list": job_list,
        "mem": mem,
        "proc": proc,
        "temp_path": temp_path
    }

    logger.info("Starting in runner mode")
    asyncio.get_event_loop().run_until_complete(virtool.jobs.run.run(config, virtool.jobs.runner.JobRunner))


@cli.command("agent", help="Start an agent that runs multiple jobs")
@click.option(
    "--job-list", "-l",
    default=["jobs_lg", "jobs_sm"],
    help="A Redis list key to pull job IDs from",
    multiple=True
)
@click.option(
    "--lg-mem",
    default=1,
    help="The maximum memory (GB) the runner may use"
)
@click.option(
    "--lg-proc",
    default=1,
    help="The maximum memory (GB) the runner may use"
)
@click.option(
    "--mem",
    default=1,
    help="The maximum memory (GB) the instance may use"
)
@click.option(
    "--proc",
    default=1,
    help="The maximum number of processes the instance can use"
)
@click.option(
    "--sm-mem",
    default=1,
    help="The maximum memory (GB) the runner may use"
)
@click.option(
    "--sm-proc",
    default=1,
    help="The maximum memory (GB) the runner may use"
)
@click.option(
    "--temp-path",
    type=click.Path(),
    help="The path to local directory for temporary files"
)
@click.pass_context
def start_agent(ctx, job_list, lg_mem, lg_proc, mem, proc, sm_mem, sm_proc, temp_path):
    virtool.logs.configure_base_logger(ctx.obj["dev"], ctx.obj["verbose"])

    config = {
        **ctx.obj,
        "job_list": job_list,
        "lg_mem": lg_mem,
        "lg_proc": lg_proc,
        "mem": mem,
        "proc": proc,
        "sm_mem": sm_mem,
        "sm_proc": sm_proc,
        "temp_path": temp_path
    }

    validate_limits(config)

    logger.info("Starting in agent mode")
    asyncio.get_event_loop().run_until_complete(virtool.jobs.run.run(config, virtool.jobs.runner.JobAgent))
