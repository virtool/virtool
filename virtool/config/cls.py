from dataclasses import dataclass
from pathlib import Path


@dataclass
class Config:
    db_connection_string: str
    db_name: str
    dev: bool
    openfga_host: str
    openfga_scheme: str
    openfga_store_name: str
    postgres_connection_string: str
    redis_connection_string: str
    b2c_client_id: str = None
    b2c_client_secret: str = None
    b2c_tenant: str = None
    b2c_user_flow: str = None
    base_url: str = ""
    data_path: Path = None
    fake: bool = False
    fake_path: Path = None
    force_version: str = None
    host: str = None
    no_check_files: bool = False
    no_check_db: bool = False
    no_fetching: bool = False
    no_revision_check: bool = False
    port: int = 9950
    use_b2c: bool = False
    verbose: bool = False
    sentry_dsn: str = ""
