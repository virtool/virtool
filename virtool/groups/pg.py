from sqlalchemy import Integer, String
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import mapped_column, Mapped
from virtool_core.models.enums import Permission

from virtool.pg.base import Base
from virtool.users.utils import generate_base_permissions


class SQLGroup(Base):
    __tablename__ = "groups"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    legacy_id: Mapped[str] = mapped_column(String, nullable=True, unique=True)
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)
    permissions: Mapped[dict[str, bool]] = mapped_column(JSONB, nullable=False)


def merge_group_permissions(groups: list[dict]) -> dict[str, bool]:
    """
    Return a :class:`dict` of permissions that will be inherited by a user belonging to
    all the passed ``groups``.

    :param groups: a list of group dictionaries
    :return: a dict keyed by permission names with boolean values indicating the state
    of the permission

    """
    permission_dict = generate_base_permissions()

    for p in Permission:
        for group in groups:
            try:
                if group["permissions"][p]:
                    permission_dict[p] = True
            except KeyError:
                permission_dict[p] = False

    return permission_dict
