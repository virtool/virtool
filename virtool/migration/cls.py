from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Awaitable, Callable

from virtool.migration.ctx import MigrationContext


class RevisionSource(str, Enum):
    """The source of a revision."""

    ALEMBIC = "alembic"
    VIRTOOL = "virtool"


@dataclass
class AppliedRevision:
    """A Virtool data revision that has been applied to the instance."""

    id: int
    applied_at: datetime
    created_at: datetime
    name: str
    revision: str


@dataclass
class GenericRevision:
    """Represents either a Virtool or Alembic revision."""

    alembic_downgrade: str | None
    created_at: datetime
    id: str
    name: str
    source: RevisionSource
    upgrade: Callable[[MigrationContext], Awaitable[None]]
    virtool_downgrade: str | None


class MigrationError(Exception):
    """Raised when an error occurs during a migration."""
