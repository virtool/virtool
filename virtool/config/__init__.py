from aiohttp.web_request import Request

from virtool.config.cls import Config
from virtool.types import App


def get_config_from_app(app: App) -> Config:
    """Get the application configuration object given the ``app`` object."""
    return app["config"]


def get_config_from_req(req: Request) -> Config:
    """Get the application configuration object given a request."""
    return get_config_from_app(req.app)
