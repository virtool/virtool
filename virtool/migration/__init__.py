"""Manage revisions in Alembic and other Virtool data.

* Virtool revisions are performed on PostgreSQL MongoDB, OpenFGA, and data files.
* Alembic revisions are for PostgreSQL only.

When should you use Virtool revisions for Postgres? When you need to make mass changes
to the database involving multiple transactions. For example, migrating OTU data from
MongoDB to PostgreSQL.

In Virtool, Alembic and non-Alembic revisions are managed together. If your newest
revision is an Alembic revision, and you create a new one. It will directly depend on
the previous Alembic revision having been applied.

# Guidelines

## MongoDB Transactions

In MongoDB, transactions fail frequently. We should avoid them for all but the smallest
operations.

It is not possible to guarantee that the revision will be recorded as complete even if
the ``upgrade()`` method succeeds. In this case, the revision will be run again next
time the migration is run.

Only write revisions for MongoDB that can be run again without causing problems and can
be re-run in the case that a previous run did not complete. You can do this by taking
the following precautions.

1. If you are adding a field, only add it to documents that do not have it.
2. If you are **not** using a transaction with ``update_many``, manually check that the
   update was successful. The ``update_many`` operation is not atomic.

   See existing revisions for how to do this.

## File Operations

Revisions containing file operations musts be written with the assumption that they
will fail and have to be re-run.

"""

from virtool.migration.cls import MigrationContext, MigrationError

__all__ = [
    "MigrationContext",
    "MigrationError",
]
