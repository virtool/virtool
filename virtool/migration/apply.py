import asyncio
import os
from pathlib import Path

import alembic.command
import alembic.config
import arrow
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession, AsyncEngine
from structlog import get_logger

from virtool.config.cls import MigrationConfig
from virtool.migration.cls import RevisionSource, GenericRevision
from virtool.migration.ctx import MigrationContext, create_migration_context
from virtool.migration.pg import (
    fetch_last_applied_revision,
    SQLRevision,
    list_applied_revisions,
)
from virtool.migration.show import load_all_revisions

logger = get_logger("migration")


async def apply(config: MigrationConfig):
    """
    Apply revisions up to the most recent revision provided by the Virtool release.

    The following safety measures are taken:
    * Revisions that have already been successfully applied will not be reapplied.
    * When a revision fails to apply, the entire migration process will stop.
    * Revisions will start applied after the last successfully applied revision.

    :param config: the configuration values for migration
    """
    os.environ["SQLALCHEMY_URL"] = config.postgres_connection_string

    all_revisions = load_all_revisions()

    for revision in all_revisions:
        logger.info(
            "Loaded revision",
            id=revision.id,
            name=revision.name,
            source=revision.source.value,
        )

    ctx = await create_migration_context(config)

    await ensure_revisions_table(ctx.pg)

    last_applied_revision = await fetch_last_applied_revision(ctx.pg)

    if last_applied_revision:
        logger.info(
            "Found last applied revision",
            revision=last_applied_revision.revision,
            name=last_applied_revision.name,
        )
    else:
        logger.info("No revisions have been applied yet")

    applied_revision_ids = {
        revision.revision for revision in await list_applied_revisions(ctx.pg)
    }

    start_applying = last_applied_revision is None

    for revision in all_revisions:
        if not start_applying:
            if revision.id == last_applied_revision.revision:
                start_applying = True

            continue

        logger.info("Checking revision", id=revision.id, name=revision.name)

        # This is necessary because buggy versions of the migration may have applied
        # revisions in the wrong order.
        if revision.id in applied_revision_ids:
            logger.info(
                "Revision is already applied", id=revision.id, name=revision.name
            )
            continue

        await apply_one_revision(ctx, revision)

    os.environ["SQLALCHEMY_URL"] = ""

    logger.info("Migration complete")


async def apply_one_revision(ctx: MigrationContext, revision: GenericRevision):
    """
    Apply a single revision to Virtool data sources.

    The revision can be either a Virtool revision or an Alembic revision. Alembic
    revisions will be applied using the Alembic CLI. Virtool revisions will be applied
    using our custom migration system.

    :param ctx: the migration context
    :param revision: the revision to apply
    """
    log = logger.bind(id=revision.id, name=revision.name)

    log.info("Applying revision")

    if revision.source == RevisionSource.ALEMBIC:
        await apply_alembic(revision.id)
    else:
        await revision.upgrade(ctx)

    async with AsyncSession(ctx.pg) as session:
        session.add(
            SQLRevision(
                applied_at=arrow.utcnow().naive,
                created_at=revision.created_at,
                name=revision.name,
                revision=revision.id,
            )
        )

        await session.commit()

    log.info("Applied revision")


async def apply_alembic(revision: str):
    """
    Apply the Alembic revision with the given id.

    """
    await asyncio.to_thread(
        alembic.command.upgrade,
        alembic.config.Config(Path(__file__).parent.parent.parent / "alembic.ini"),
        revision,
        False,
        None,
    )


async def ensure_revisions_table(pg: AsyncEngine):
    """
    Ensure that the `revisions` table exists in the database.

    :param pg: the PostgreSQL database connection
    """

    async with AsyncSession(pg) as session:
        async with session.begin():
            await session.execute(
                text(
                    """
                    CREATE TABLE IF NOT EXISTS revisions (
                        id SERIAL PRIMARY KEY,
                        name varchar(64) NOT NULL,
                        revision varchar(18) NOT NULL,
                        created_at timestamp without time zone NOT NULL,
                        applied_at timestamp without time zone NOT NULL
                    )
                    """
                )
            )
