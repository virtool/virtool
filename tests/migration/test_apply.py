import datetime
import shutil
from dataclasses import asdict
from pathlib import Path

import pytest
from sqlalchemy.ext.asyncio import AsyncEngine
from syrupy.matchers import path_type

from virtool.config.cls import MigrationConfig
from virtool.migration.pg import list_applied_revisions
from virtool.migration.apply import apply


@pytest.fixture
def example_revisions_path(revisions_path: Path):
    shutil.copytree(Path(__file__).parent / "revisions", revisions_path)


async def test_apply_revisions(
    example_revisions_path: Path,
    migration_config: MigrationConfig,
    pg: AsyncEngine,
    snapshot,
):
    await apply(migration_config)

    assert [asdict(r) for r in await list_applied_revisions(pg)] == snapshot(
        matcher=path_type(
            {
                ".*applied_at": (datetime.datetime,),
                ".*created_at": (datetime.datetime,),
            },
            regex=True,
        )
    )
