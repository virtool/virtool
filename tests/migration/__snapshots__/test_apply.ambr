# serializer version: 1
# name: test_apply_revisions
  list([
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 1,
      'name': 'Add missing schemas',
      'revision': 'e694fb270acb',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 2,
      'name': 'upgrade file size types',
      'revision': '90bf491700cb',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 3,
      'name': 'Nest analysis results field',
      'revision': '7emq1brv0zz6',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 4,
      'name': 'Add subtractions field',
      'revision': '1keyha5n6l0j',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 5,
      'name': 'Rename cache hash field to key',
      'revision': 'jhqn47cauoea',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 6,
      'name': 'Remove ref process field',
      'revision': '1p681ke9wedv',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 7,
      'name': 'Remove software_update status document',
      'revision': 'gr5j6jx8ap7f',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 8,
      'name': 'Add user handle',
      'revision': 'ydvidjp34n4c',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 9,
      'name': 'Add user active field',
      'revision': 'tlogeiyxl9uz',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 10,
      'name': 'Use null for unset analysis results fields',
      'revision': 'i0ljixkr0wxg',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 11,
      'name': 'Add jobs ping field',
      'revision': '6q5k8tz8uph3',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 12,
      'name': 'Create all database indexes',
      'revision': 'l20h8fsbbb28',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 13,
      'name': 'Add space field',
      'revision': 's607ucxpct81',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 14,
      'name': 'Add space table and columns',
      'revision': '8af76adc9706',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 15,
      'name': 'Populate sample workflows field',
      'revision': '1zg28cpib2uj',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 16,
      'name': 'Add groups table',
      'revision': '7bd75961d130',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 17,
      'name': 'Fix group permission type',
      'revision': '011389a5ae19',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 18,
      'name': 'Create ML model tables.',
      'revision': '814389ec5021',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 19,
      'name': 'Move groups to SQL',
      'revision': 'r4txubh413sp',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 20,
      'name': 'Add ML model release size',
      'revision': 'f8ad70032e9c',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 21,
      'name': 'Use integer nested group ids',
      'revision': 'okd9db53g0il',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 22,
      'name': 'Use integer group ids in users',
      'revision': 'f4md6q93mtud',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 23,
      'name': 'Use integer group ids in samples',
      'revision': 'qg7bo42yvavg',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 24,
      'name': 'add users table',
      'revision': '4104346698df',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 25,
      'name': 'rename users table primary_group to primary_group_id',
      'revision': '8f3810c1c2c9',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 26,
      'name': 'promote mongo administrators in openFGA',
      'revision': 'xtbq4bjlpocu',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 27,
      'name': 'repair schema',
      'revision': '77be1d95da09',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 28,
      'name': 'update user schema',
      'revision': '5df816cef12f',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime,
      'id': 29,
      'name': 'add user email',
      'revision': '141c7ecb99b7',
    }),
  ])
# ---
